### DISCUSSION TITLE
Signals as the "primary vehicle" for user-interaction

### DISCUSSION THREAD

**Discussion by bzm3r** [Leafwing-Studios/Emergence#60] [Design notes]
In biology, signals are the life-blood of organization. See for instance the classic [Sniffers, Buzzers, Toggles and Blinkers](https://jigcell.cs.vt.edu/Pubs/Tyson03.pdf). Don't let the particular scale of biology in that article faze you: signals are relevant regardless of scale, although chemical signalling are especially relevant at the scale of our game, and smaller. An awesome book that is relevant: [An Introduction to Systems Biology by Uri Alon](https://www.routledge.com/An-Introduction-to-Systems-Biology-Design-Principles-of-Biological-Circuits/Alon/p/book/9781439837177) (I have read the introduction to this, I should really read the rest of it...but anxiety!)

For the purposes of discussion we lean on the chemical side of things, but both reader and author should keep a mind out for how this abstracts beyond chemistry.
 
In part, the awesomeness of signals comes from the ability of organisms to design signals that interact with other signals (through chemistry). For example, signals A, B, C can react in any of the following ways (hint: signal interaction be represented/abstracted through a simple algebra):
* A + B + time = C, or 
* A + B + time = A, or 
* A + B + time = 2A
* A + time = B 

The other part of the awesomeness of signals comes form the ability of organisms to design the time characteristics of signals; for example:
* how long-lived signals are (their half-life)
* how quickly/slowly they react with other signals
* how quickly/slowly they diffuse or are transported (this is where the connection to space (not rocket space, but living space) comes in)

By playing around with these interaction and time, organisms can achieve:
* patterning (zoning)
* communication of information
* dynamic response (change in behaviour given change in conditions)

Therefore, an alternative take on interaction could be the following:

* provide an editor/environment for users to specify relationships between signals, and the time characteristics of signals
* provide an editor/environment for users to specify how a signal receiver combines input signals to produce an output signal (or response of other sort)

(So, at the very least this line of thinking also nudges one to consider pheromones as *the* way by which zoning is achieved. Put differently, zoning actions should reduce the pheromone bar, and zoning actions should not be permanent but should achieve longevity through other re-enforcers?)

**Comment by alice-i-cecile** [comment]
> So, at the very least this line of thinking also nudges one to consider pheromones as the way by which zoning is achieved. Put differently, zoning actions should reduce the pheromone bar, and zoning actions should not be permanent but should achieve longevity through other re-enforcers

Definitely think that changing zoning should deplete the pheromone bar. It's effectively a mechanic to deter micromanagement :)

Unsure about "should this be permanent". I worry a lot about this creating unnecessary busy work, or a compulsion to place down reinforcers everywhere. The other thing I've considered is letting the player boost zoning strength temporarily as a pheromone type. I think this gives a similar effect, without the frustration.

**Reply by bzm3r** [reply]
Yeah, I think your comment helped me realize something: the zoning mechanic essentially is for the player to help manage their planning; it can be viewed separately from the biology/physics/chemistry happening on-the-ground. It's more like the mental picture the hive-mind has.

**Reply by alice-i-cecile** [reply]
Yep, that's exactly how I would describe it :) You'll likely want to be able to e.g. copy-and-paste zoning configurations down the line.

**Comment by alice-i-cecile** [comment]
> provide an editor/environment for users to specify relationships between signals, and the time characteristics of signals
> provide an editor/environment for users to specify how a signal receiver combines input signals to produce an output signal (or response of other sort)

Yep, I've been toying with this idea in the past. I think this makes most sense as something that can be calibrated at the species level. This provides incentives for speciation and hooks for specialization.

You want to attach some cost to messing with these settings (even just a lockout period for each setting) to prevent players using this to micro.

**Comment by bzm3r** [comment]
One possible idea for what the signalling editor could roughly look like (from SpaceChem):

![image](https://user-images.githubusercontent.com/4350451/194722539-039771fa-32e4-4071-8056-5a77544da5cc.png)

Key point: it's a kind of visual programming tool, that also includes notions of timing.
